doc/source/data/performance-tips.rst

```diff
@@ -25,17 +25,122 @@ Tuning read parallelism
 
 By default, Ray Data automatically selects the read ``parallelism`` according to the following procedure:
 
-1. The number of available CPUs is estimated. If in a placement group, the number of CPUs in the cluster is scaled by the size of the placement group compared to the cluster size. If not in a placement group, this is the number of CPUs in the cluster.
-2. The parallelism is set to the estimated number of CPUs multiplied by 2. If the parallelism is less than 8, it's set to 8.
-3. The in-memory data size is estimated. If the parallelism would create in-memory blocks that are larger on average than the target block size (512MiB), the parallelism is increased until the blocks are < 512MiB in size.
+The ``parallelism`` parameter passed to Ray Data's :ref:`read APIs <input-output>` specifies the number of read tasks that should be used.
+Usually, if the read is followed by a :func:`~ray.data.Dataset.map` or :func:`~ray.data.Dataset.map_batches`, the map is fused with the read; therefore ``parallelism`` also determines the parallelism used during map.
 
-Occasionally, it's advantageous to manually tune the parallelism to optimize the application. You can do this by setting the ``parallelism`` parameter.
-For example, use ``ray.data.read_parquet(path, parallelism=1000)`` to force up to 1000 read tasks to be created.
+The default value for ``parallelism`` is decided based on the following heuristics, applied in order:
+
+1. We start with the default parallelism of 200. This can be overridden by setting :class:`DataContext.min_parallelism <ray.data.context.DataContext>`.
+2. Min block size (default=1MiB). If the parallelism would make blocks smaller than this threshold, the parallelism is reduced to avoid the overhead of tiny blocks. This can be overridden by setting :class:`DataContext.target_min_block_size <ray.data.context.DataContext>` (bytes).
+3. Max block size (default=128MiB). If the parallelism would make blocks larger than this threshold, the parallelism is increased to avoid OOMs during processing. This can be overridden by setting :class:`DataContext.target_max_block_size <ray.data.context.DataContext>` (bytes).
+4. Available CPUs. If the parallelism cannot make use of all the available CPUs in the cluster, the parallelism is increased until it can. Ray Data chooses read tasks be at least 2x the number of available CPUs.
+
+Occasionally, it's advantageous to manually tune the parallelism to optimize the application.
+For example, the following code will batch multiple files into the same read task to avoid creating blocks that are too large.
+
```

Comment:
Here and whenever else we call `ray.init`:
```suggestion

.. testcode::
    :hide:
    
    import ray
    ray.shutdown()
    
```

Reply:
Why not show it? That way the code is completely copy-pastable.

Reply:
We could. I was assuming that readers wouldn't have already run `ray.init()` when they copy the code snippet. If we want to include `ray.shutdown()`, we could just add it directly in the other `testcode`.

The documentation test runs all of the code snippets serially within a file, so if we don't do this we might get errors.